### PR TITLE
Add documentation for the `-Zself-profile` flag

### PR DESCRIPTION
**Pull request by wesleywiser**

### CODE REVIEW

**Comment by andjo403** [comment]
nice work. shall there maybe also be some short description of the -Zself_profile_events?

**Comment by wesleywiser** [comment]
@andjo403 Yeah, that's a good idea!

**Comment by andjo403** [comment]
this looks amazing

**Comment by wesleywiser** [comment]
Thanks!

**Comment by wesleywiser** [comment]
cc @michaelwoerister

File: src/doc/unstable-book/src/compiler-flags/self-profile-events.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+## Event types
+
+- `query-provider`
+  - Traces each query used internally by the compiler.
+
+- `generic-activity`
+  - Traces other parts of the compiler not covered by the query system.
+
+- `query-cache-hit`
+  - Adds tracing information that records when the query cache is "hit" and does not need to re-execute a query which has been cached.
```
Comment (michaelwoerister) [inline src/doc/unstable-book/src/compiler-flags/self-profile-events.md]: "query cache" -> "in-memory query cache"

File: src/doc/unstable-book/src/compiler-flags/self-profile-events.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  - Disabled by default because this significantly increases the trace file size.
+
+- `query-blocked`
+  - Tracks time that a query tries to run but is blocked waiting on another query to finish executing.
```
Comment (michaelwoerister) [inline src/doc/unstable-book/src/compiler-flags/self-profile-events.md]: "waiting on another query" -> "waiting on another thread that executes the same query"

 That would be more accurate.

File: src/doc/unstable-book/src/compiler-flags/self-profile-events.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  - Query blocking only occurs when the compiler is built with parallel mode support.
+
+- `incr-cache-load`
+  - Tracks time that is spent loading and deserializing query results from the incremental cache.
```
Comment (michaelwoerister) [inline src/doc/unstable-book/src/compiler-flags/self-profile-events.md]: "incremental cache" -> "incremental compilation on-disk cache"

File: src/doc/unstable-book/src/compiler-flags/self-profile-events.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+- `query-keys`
+  - Adds a serialized representation of each query's query key to the tracing data.
+  - Disabled by default because this significantly increases the trace file size.
+
+- `function-args`
+  - Adds additional tracing data to some `generic-activity` events.
+  - Disabled by default for parity with `query-keys`.
```
Comment (michaelwoerister) [inline src/doc/unstable-book/src/compiler-flags/self-profile-events.md]: off-topic: we should measure if enabling this by default adds noticeable overhead.

File: src/doc/unstable-book/src/compiler-flags/self-profile-events.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+- `none`
+  - Disables all events.
+  Equivalent to the self-profiler being disabled.
+
+- `default`
+  - The default set of events which stikes a balance between providing detailed tracing data and adding additional overhead to the compilation.
+
+- `all`
+  - Enables all events.
```
Comment (michaelwoerister) [inline src/doc/unstable-book/src/compiler-flags/self-profile-events.md]: `args` could also be added to this list.

**Comment by michaelwoerister** [comment]
Looks great! I left a few minor suggestions.

**Comment by wesleywiser** [comment]
Thanks @michaelwoerister! I applied your suggestions.

**Review by nikomatsakis — APPROVED** [review]
This is wonderful, thanks @wesleywiser

**Comment by nikomatsakis** [comment]
@bors rollup=always

Just edits documentation, low risk.

**Comment by wesleywiser** [comment]
Thanks @nikomatsakis! Once this lands, I'll update the blog post to link to these docs, update the post date and then we'll be ready to post it.

**Pull request merged**